### PR TITLE
Store app settings in a new doc

### PR DESCRIPTION
**Pull request by garethbowen**
# Description

Moves the settings from the ddoc.app_settings into a new doc with id "settings". This means the settings are maintained when deploying a new app version and don't need to be copied to the medic-client ddoc to get replicated to the client.

medic/medic-webapp#4541

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

File: api/src/controllers/changes.js [hunk 1]
```diff
@@ -391,6 +391,7 @@ var getReplicationKey = function(doc) {
   if (doc._id === 'resources' ||
```
Comment (SCdF) [inline api/src/controllers/changes.js]: the WARNING comment about this points to the wrong place now

File: api/src/ddoc-extraction.js [hunk 2]
```diff
@@ -59,6 +56,9 @@ const isUpdated = (settings, newDdoc) => {
         oldDdoc._attachments = newDdoc._attachments;
       }
 
+      // delete the obsolete app_settings so the docs will be comparable
+      delete oldDdoc.app_settings;
```
Comment (SCdF) [inline api/src/ddoc-extraction.js]: would it be crazy to do this in a migration and then just presume we're fine at this stage?

File: api/src/ddoc-extraction.js [hunk 3]
```diff
@@ -105,9 +105,30 @@ const findUpdatedAppcache = ddoc => {
     });
 };
 
+// converts old style app_settings on ddoc to new separate doc
+const extractAppSettings = ddoc => {
```
Comment (SCdF) [inline api/src/ddoc-extraction.js]: again, if we do a migration can we just not have this code at all?

File: sentinel/src/config.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ const initFeed = () => {
   // Use since=now on ddoc listener so we don't replay an old change.
   const feed = new follow.Feed({ db: process.env.COUCH_URL, since: 'now' });
   feed.on('change', change => {
-    if (change.id === '_design/medic') {
+    if (change.id === 'settings') {
```
Comment (SCdF) [inline sentinel/src/config.js]: is now the time to create a SETTINGS SHARED LIBRARY!?!?!?!?

Hyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyype

File: webapp/src/ddocs/medic/validate_doc_update.js [hunk 5]
```diff
@@ -10,7 +10,7 @@
 function(newDoc, oldDoc, userCtx, secObj) {
 
   var ADMIN_ONLY_TYPES = [ 'form', 'translations' ],
-      ADMIN_ONLY_IDS = [ 'resources', 'appcache', 'zscore-charts' ];
+      ADMIN_ONLY_IDS = [ 'resources', 'appcache', 'zscore-charts', 'settings' ];
```
Comment (SCdF) [inline webapp/src/ddocs/medic/validate_doc_update.js]: 👍

File: api/src/ddoc-extraction.js (hunk 2 above)
Comment (garethbowen) [reply api/src/ddoc-extraction.js]: ddoc extraction runs before migrations on startup, and then after migrations on change, so we can't assume it's fine at this stage. However I think if it's implemented as a migration we don't need to do it here.

File: sentinel/src/config.js (hunk 4 above)
Comment (garethbowen) [reply sentinel/src/config.js]: Yeah maybe. I'll create a todo to look at if after this is merged.

File: api/src/controllers/changes.js (hunk 1 above)
Comment (garethbowen) [reply api/src/controllers/changes.js]: Fixed.

**Comment by garethbowen** [comment]
@SCdF You know what? So long as everyone upgrades via Horti their settings will be migrated before api starts, so why do we need code in api to handle it either in ddoc-extraction _or_ a migration? Developers are a special case but they can run the manual script to keep their settings if required. What do you think? Is there an edge case I'm forgetting?

**Comment by SCdF** [comment]
@garethbowen I think you are right here. Devs will only lose their settings just once, when they pull this merged change. And they can run the script after they notice they've lost it right, because it will still be on the ddoc?

How are we migrating 2.x -> 3.0? I'm wondering if we're covering that case. If we:
 - move their data onto a server with horti (or somehow contort their server into having horti on it and remove gardener)
 - manually run `horti --bootstrap=3.0.0`

Then I think this code is fine. As long as horti runs before the code does, which I think it has to (ie we're not doing some kind of weird docker _in situ_ thing where it already has the code baked in or something)

**Pull request merged**